OSDOCS-4152: Installing a private cluster on IBM Cloud VPC

mjpytlak opened the pull request:
Version(s):
4.12+

Issue:
This issue addresses [OSDOCS-4152](https://issues.redhat.com/browse/OSDOCS-4152)

Link to docs preview:

- Configuring DNS resolution > [Using IBM Cloud DNS Services for DNS Resolution](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-account.html#installation-dns-ibm-cloud_installing-ibm-cloud-account)
- IBM Cloud VPC IAM Policies and API Key > [Required access policies](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-account.html#required-access-policies-ibm-cloud_installing-ibm-cloud-account)
- [Private clusters](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private)
- [Private clusters in IBM Cloud VPC](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private)
- [Obtaining the installation program](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-obtaining-installer_installing-ibm-cloud-private)
- [Manually creating the installation configuration file](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-initializing-manual_installing-ibm-cloud-private)
- Installation configuration parameters > [Network configuration parameters](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-configuration-parameters-network_installing-ibm-cloud-private)
- Installation configuration parameters > [Additional IBM Cloud VPC configuration parameters](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-configuration-parameters-additional-ibm-cloud_installing-ibm-cloud-private)
- [Sample customized install-config.yaml file for IBM Cloud VPC](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-ibm-cloud-config-yaml_installing-ibm-cloud-private)

mjpytlak commented:
The following sections are specific to private installations. At a minimum, this content will require updates. Please review and comment on what needs to be added/updated/removed:

- [Private clusters](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private) (modules/private-clusters-default.adoc) This is boiler plate content in all of the installation topics for private installations. Highly doubt we need to update this, but thought it worth calling out.

jeffnowicki reviewed (commented):
Some initial comments.

- we need to refactor/enhance the prereq for DNS resolution configuration (public will use CIS service and private will use DNS service) - can discuss at our review call
- private install requires an existing vpc - unfortunately, the docs in this PR didn't reflect the updates in the existiing vpc PR - https://github.com/openshift/openshift-docs/pull/50357. i attempted to point out a few of the obvious ones.
- otherwise, I think this looks reasonably good.
- look forward to re-reviewing docs after both PRs (existing vpc, private cluster) have merged.

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-private.adoc:
```diff
@@ -0,0 +1,66 @@
+* You read the documentation on xref:../../installing/installing-preparing.adoc#installing-preparing[selecting a cluster installation method and preparing it for users].
+* You xref:../../installing/installing_ibm_cloud_public/installing-ibm-cloud-account.adoc#installing-ibm-cloud-account[configured an IBM Cloud account] to host the cluster.
+* If you use a firewall, you xref:../../installing/install_config/configuring-firewall.adoc#configuring-firewall[configured it to allow the sites] that your cluster requires access to.
+* You configured the `ccoctl` utility before you installed the cluster. For more information, see xref:../../installing/installing_ibm_cloud_public/configuring-iam-ibm-cloud.adoc#configuring-iam-ibm-cloud[Configuring IAM for IBM Cloud VPC].
```
Correct. I agree that revising the content will be more user friendly. Given that the current content is technically accurate, this update is beyond the scope of this PR and a separate docs issue will be created to track the update through completion, which I still need to create. Thank you for the reminder.

mjpytlak commented on modules/installation-ibm-cloud-config-yaml.adoc:
```diff
@@ -21,6 +29,7 @@ You can customize the `install-config.yaml` file to specify more details about y
 This sample YAML file is provided for reference only. You must obtain your `install-config.yaml` file by using the installation program and modify it.
```
All set.

mjpytlak commented on modules/installation-configuration-parameters.adoc:
```diff
@@ -630,13 +638,13 @@ endif::openshift-origin[]
+Setting this field to `Internal` is not supported on non-cloud platforms.
 ifeval::[{product-version} <= 4.7]
 [IMPORTANT]
 ====
```
Done. Thank you.

mjpytlak commented on modules/installation-obtaining-installer.adoc:
```diff
@@ -24,6 +24,8 @@
 // * installing/installing_ibm_cloud_public/installing-ibm-cloud-customizations.adoc
 // * installing/installing_ibm_cloud_public/installing-ibm-cloud-network-customizations.adoc
+// * installing/installing_ibm_cloud_public/installing-ibm-cloud-vpc.adoc
+// * installing/installing_ibm_cloud_public/installing-ibm-cloud-private.adoc
```
This is a good catch. I updated the intro to "Obtaining the installation program" to align with the installation reqs for a private cluster.

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-private.adoc:
```diff
@@ -0,0 +1,66 @@
+
+* You reviewed details about the xref:../../architecture/architecture-installation.adoc#architecture-installation[{product-title} installation and update] processes.
+* You read the documentation on xref:../../installing/installing-preparing.adoc#installing-preparing[selecting a cluster installation method and preparing it for users].
+* You xref:../../installing/installing_ibm_cloud_public/installing-ibm-cloud-account.adoc#installing-ibm-cloud-account[configured an IBM Cloud account] to host the cluster.
```
Thank you for highlighting this. I had vaguely recalled a conversation in Slack [1] about private installs and using DNS Services. Based on that info, I went ahead and broke the content around configuring DNS resolution into a section for using IBM Cloud Internet Services and a section for using IBM Cloud DNS Services.

[1] https://coreos.slack.com/archives/C01U40AM37F/p1663618418824469?thread_ts=1663618379.260929&cid=C01U40AM37F

jeffnowicki commented on modules/private-clusters-about-ibm-cloud.adoc:
```diff
@@ -0,0 +1,24 @@
+[id="private-clusters-limitations-ibm-cloud_{context}"]
+== Limitations
+
+<TBD>
```
Let's start with this for limitations (following what Azure did). Will discuss further with Chris and discuss from a user/admin perspective.

"Private clusters on IBM Cloud VPC are subject only to the limitations associated with the existing VPC that was used for cluster deployment."

jeffnowicki commented:
@mjpytlak @mjpytlak I noticed the 'choose a method' section is missing both 'installing into an existing VPC' and 'installing a private cluster'. There should now be 4 IPI methods in this section.

https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/preparing-to-install-on-ibm-cloud.html#choosing-an-method-to-install-ocp-on-ibm-cloud-installer-provisioned

mjpytlak commented:
Greetings @MayXuQQ and @jeffnowicki. I have added the approved VPC content from https://github.com/openshift/openshift-docs/pull/50357, which rounds out the doc for installing to a private cluster.

@jeffnowicki To date I have addressed all of your feedback. Please take one more pass.
@MayXuQQ This is ready for QE review. PTAL. Thanks!

MayXuQQ commented on modules/installation-cis-ibm-cloud.adoc:
```diff
@@ -4,9 +4,9 @@
 :_content-type: PROCEDURE
 [id="installation-cis-ibm-cloud_{context}"]
-= Configuring DNS resolution using Cloud Internet Services
+= Using IBM Cloud Internet Services for DNS resolution
```
DNS => CIS ?

mjpytlak replied:
Note the section above (Configuring DNS resolution). It explains that depending on the type of cluster you are installing determines how you configure DNS resolutions. The steps for CIS are directly below the section you are referencing. [1]

[1] [Using IBM Cloud DNS Services for DNS resolution](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-account.html#installation-dns-ibm-cloud_installing-ibm-cloud-account)

MayXuQQ replied:
How about change the title as `Using IBM Cloud Internet Services (CIS) for DNS resolution`

MayXuQQ commented on modules/installation-dns-ibm-cloud.adoc:
```diff
@@ -0,0 +1,66 @@
+= Using IBM Cloud DNS Services for DNS resolution
+
+The installation program uses IBM Cloud DNS Services to configure cluster DNS resolution and provide name lookup for a private cluster.
+
```
its is the baseDomain value in install-config.yaml

mjpytlak replied:
I am not sure what you mean by this. Please explain a bit more.

MayXuQQ replied:
yes

mjpytlak replied:
Thank you. Done.

MayXuQQ commented:
https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private
To deploy a private cluster, you must:
prepare the the dns zone as the baseDoman, you can share it with other clusters.

MayXuQQ commented:
https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-configuration-parameters-network_installing-ibm-cloud-private
Table 2. Network parameters
networking.machineNetwork.cidr need cover the cidr of the subnets of the vpc

MayXuQQ commented on modules/installation-ibm-cloud-config-yaml.adoc:
```diff
@@ -203,6 +207,101 @@ For production {product-title} clusters on which you want to perform installatio
+  - cidr: 10.128.0.0/14 <5>
+    hostPrefix: 23
+  machineNetwork:
+  - cidr: 10.0.0.0/16
```
machineNetwork cidr need cover the cidrs of the subnets in the vpc.

mjpytlak replied:
Done. I noticed that we do not call out this fact in the documentation for installing to an existing VPC. Should I update that sample as well?

[1] https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-vpc.html#installation-ibm-cloud-config-yaml_installing-ibm-cloud-vpc

MayXuQQ replied:
install private cluster based on install on an existing VPC, baseDomain is different based on publish value.
the private cluster fail to prompt the baseDomain value.

MayXuQQ commented on modules/installation-ibm-cloud-config-yaml.adoc:
```diff
@@ -203,6 +207,101 @@ For production {product-title} clusters on which you want to perform installatio
+  additionalEnabledCapabilities:
+  - openshift-samples
+----
+<1> Required. The installation program prompts you for this value.
```
for private cluster use the dns zone, not prompt by installation program. 
$ ic dns instances
Name                             ID                                     Location   State    Service Name   
preserved-openshift-qe-private   54cc61f8-28bc-4f9a-a815-e07dfe705fed   global     active   dns-svcs   
$ ic dns zones 
Listing zones for service instance 'preserved-openshift-qe-private' ...
ID                                     Name                                           Status   
3c7af30d-cc2c-4abc-94e1-3bcb36e01a9b   **private-ibmcloud.qe.devcluster.openshift.com** PENDING_NETWORK_ADD

mjpytlak replied:
This is an excellent catch. I had completely overlooked the fact that you cannot use the installation program to create the the `install-config.yaml` file. You must, in fact, create it manually. I have gone ahead and added the correct procedure for doing that.

MayXuQQ commented:
hide https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-configure-proxy_installing-ibm-cloud-private 
now private cluster behind proxy has bug [OCPBUGS-2363](https://issues.redhat.com/browse/OCPBUGS-2363)

MayXuQQ commented:
https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#manually-create-iam-ibm-cloud_installing-ibm-cloud-private
4.path_to_credential_requests_directory  and 5 path_to_store_credential_request_templates is the same directory, how about use same name ?

mjpytlak commented:
> https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private To deploy a private cluster, you must: prepare the the dns zone as the baseDoman, you can share it with other clusters.

mjpytlak commented:
> https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private To deploy a private cluster, you must: prepare the the dns zone as the baseDoman, you can share it with other clusters.

I add a statement about needing to create the DNS zone, but it is unclear to me why we would need to elaborate further by saying that you can share it with other clusters. Is the reasoning that the DNS zone can be shared with other public clusters, but in the case of a private cluster, the installation program uses the `baseDomain` to create a private DNS zone and the required records for the cluster. If so, the last bullet in this topic [1] about needing a public DNS zone seems to conflict with what we are trying to say. Should I remove it?

[1] https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-about-ibm-cloud_installing-ibm-cloud-private

mjpytlak commented:
> https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private To deploy a private cluster, you must: prepare the the dns zone as the baseDoman, you can share it with other clusters.

Thank you. Done.

MayXuQQ commented:
> > https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private To deploy a private cluster, you must: prepare the the dns zone as the baseDoman, you can share it with other clusters.
> 
> I add a statement about needing to create the DNS zone, but it is unclear to me why we would need to elaborate further by saying that you can share it with other clusters. Is the reasoning that the DNS zone can be shared with other public clusters, but in the case of a private cluster, the installation program uses the `baseDomain` to create a private DNS zone and the required records for the cluster.

the DNS zone can be shared with other private cluster as the baseDomain in the install-config.yaml.

MayXuQQ commented on modules/private-clusters-about-ibm-cloud.adoc:
```diff
@@ -0,0 +1,24 @@
+* A public DNS zone that matches the `baseDomain` for the cluster
+
+The installation program does use the `baseDomain` that you specify to create a private DNS zone and the required records for the cluster. The cluster is configured so that the Operators do not create public records for the cluster and all cluster machines are placed in the private subnets that you specify.
+
```
1. the DSN zone need be prepared before.
2.  installation program just bind with the network  of VPC which specified in install-config.

mjpytlak replied:
Correct. We now clearly state the need for users to create a DNS zone in:

- Configuring DNS resolution > [Using IBM Cloud DNS Services for DNS resolution](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-account.html#installation-dns-ibm-cloud_installing-ibm-cloud-account)
- [Private clusters](https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#private-clusters-default_installing-ibm-cloud-private)

mjpytlak commented:
> hide https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-configure-proxy_installing-ibm-cloud-private now private cluster behind proxy has bug [OCPBUGS-2363](https://issues.redhat.com/browse/OCPBUGS-2363)

This appears to block the delivery of a disconnected cluster, not a private cluster.

mjpytlak commented:
> https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#manually-create-iam-ibm-cloud_installing-ibm-cloud-private 4.path_to_credential_requests_directory and 5 path_to_store_credential_request_templates is the same directory, how about use same name ?

Is it technically accurate to use the same name @MayXuQQ? This is not content that I own, but share with another team.
CC: @jeana-redhat

mjpytlak commented:
> https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-configuration-parameters-network_installing-ibm-cloud-private Table 2. Network parameters networking.machineNetwork.cidr need cover the cidr of the subnets of the vpc

I believe the statement I added addresses you comment. Please let me know what you think.

MayXuQQ commented:
> OCPBUGS-2363

OCPBUGS-2363 bug blocked disconnect private cluster, 
private cluster is OK. 
proxy + public cluster is OK

MayXuQQ commented:
> > https://50561--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-configuration-parameters-network_installing-ibm-cloud-private Table 2. Network parameters networking.machineNetwork.cidr need cover the cidr of the subnets of the vpc
> 
> I believe the statement I added addresses you comment. Please let me know what you think.
it is OK

MayXuQQ commented:
/lgtm

jianping-shu commented:
@mjpytlak I think these two dictionaries in step 4 & 5 are the same and can be updated as proposal in theory. Need @jeana-redhat confirm there's no side effect.

jeana-redhat commented:
> @mjpytlak I think these two dictionaries in step 4 & 5 are the same and can be updated as proposal in theory. Need @jeana-redhat confirm there's no side effect.

I can't think of any problems this creates for the content, and since both QA folks agree that this directory is the same from a functional perspective, I think that sounds good to me :+1:

jeffnowicki commented:
Nice work @mjpytlak

/lgtm

jeffnowicki commented on modules/installation-ibm-cloud-config-yaml.adoc:
```diff
@@ -203,6 +207,102 @@ For production {product-title} clusters on which you want to perform installatio
+<1> Required.
+<2> If you do not provide these parameters and values, the installation program provides the default value.
+<3> The `controlPlane` section is a single mapping, but the `compute` section is a sequence of mappings. To meet the requirements of the different data structures, the first line of the `compute` section must begin with a hyphen, `-`, and the first line of the `controlPlane` section must not. Only one control plane pool is used.
+<4> Enables or disables simultaneous multithreading, or `hyperthreading`. By default, simultaneous multithreading is enabled to increase the performance of your machines' cores. You can disable it by setting the parameter value to `Disabled`. If you disable simultaneous multithreading in some cluster machines, you must disable it in all cluster machines.
```
You can go with the 'intel implementation' version.

mjpytlak replied:
Good catch. Not sure why this content originally listed hyperthreading as a literal.

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-account.adoc:
```diff
@@ -15,7 +15,15 @@ Before you can install {product-title}, you must configure an IBM Cloud account.
 include::modules/quotas-and-limits-ibm-cloud.adoc[leveloffset=+1]
 
-include::modules/installation-cis-ibm-cloud.adoc[leveloffset=+1]
+== Configuring DNS resolution
```
Thank you. Done.

mjpytlak replied:
Yep. Thanks.

mjpytlak commented:
> the compute is ?
> 
> Compute | RHCOS, RHEL 8.4, or RHEL 8.5 [3] | 2 | 8 GB | 100 GB | 300

@jeffnowicki told me that the values for all three rows where RHCOS | 4 | 16 | 100 | 300. Jeff, please confirm.

MayXuQQ commented:
/lgtm

mjpytlak commented:
/cherrypick enterprise-4.12
